### PR TITLE
chore: replace mongodb-extended-json with bson

### LINKED ISSUE #107: `mongodb-extended-json` is deprecated
`mongodb-extended-json` is [deprecated](https://www.npmjs.com/package/mongodb-extended-json) and must be replaced with [bson](https://www.npmjs.com/package/bson).

## Changes (draft)
_/lib/index.js_
```diff
- const EJSON = require('mongodb-extended-json');
+ const { EJSON } = require('bson');
```

### PR DESCRIPTION
**Pull request by rtritto**
Fix #107

## Open points with changes:
- [x] Date, ISODate => no ms with 000 ms ([ref](https://jira.mongodb.org/browse/NODE-5030))
- [x] BinData => added 0 prefix to _$type_
- [x] UUID => added 0 prefix to _$type_
- [x] Timestamp => added _i_ and _t_ properties
- [x] NumberInt => added _$numberInt_

### CODE REVIEW

**Comment by addaleax** [comment]
@rtritto Hi! Just a heads up, in order for us to accept contributions, we need a signed CLA: https://www.mongodb.com/contributor

**Comment by rtritto** [comment]
> @rtritto Hi! Just a heads up, in order for us to accept contributions, we need a signed CLA: https://www.mongodb.com/contributor

After login I got redirected to https://cloud.mongodb.com, but when I go to your link page, there is no authentication.

**Comment by rtritto** [comment]
I don't know how to fix _NumberLong_ tests.

PS: I updated the README

**Comment by rtritto** [comment]
@addaleax I used `relaxed=false` and fixed tests.
Can you review?

File: lib/index.js [hunk 1]
```diff
@@ -93,7 +93,10 @@ module.exports.isFilterValid = (input, options = {}) => {
     const parsed = parseFilter(input);
     if (options.validate === false) return parsed;
     // is it a valid MongoDB query according to the language?
-    return queryLanguage.accepts(EJSON.stringify(parsed)) ? parsed : false;
+    return queryLanguage.accepts(JSON.stringify(EJSON.serialize(parsed, {
```
Comment (addaleax) [inline lib/index.js:96]: How is this different from `EJSON.stringify(parsed, ...)`, though? `EJSON.serialize` is implemented as `JSON.parse(EJSON.stringify(...))`, so we should be able to skip the extra serialization/deserialization round-trip:

https://github.com/mongodb/js-bson/blob/6ef100083328b2eeeb3c6a536985f494f919071f/src/extended_json.ts#L464

File: README.md [hunk 2]
```diff
@@ -33,12 +33,12 @@ parser.toJavascriptString(parse(query));
...
 
 ```javascript
+var bson = require('bson');
 var parser = require('mongodb-query-parser');
-var EJSON = require('mongodb-extended-json');
 var queryAsAnObjectWithTypes = parser.parseFilter(query);
 
 // Use extended json to prove types are intact
-EJSON.stringify(queryAsAnObjectWithTypes);
+JSON.stringify(bson.EJSON.serialize(queryAsAnObjectWithTypes));
```
Comment (addaleax) [inline README.md:41]: (ditto here, same question as below)

**Review by addaleax — COMMENTED** [review]
> > @rtritto Hi! Just a heads up, in order for us to accept contributions, we need a signed CLA: https://www.mongodb.com/contributor

> After login I got redirected to https://cloud.mongodb.com, but when I go to your link page, there is no authentication.

Are you saying that https://www.mongodb.com/contributor (or https://www.mongodb.com/legal/contributor-agreement) redirects you to cloud.mongodb.com? I don’t think that page should require any login at all

**Comment by rtritto** [comment]
> Are you saying that https://www.mongodb.com/contributor (or https://www.mongodb.com/legal/contributor-agreement) redirects you to cloud.mongodb.com? I don’t think that page should require any login at all

I signed with success.

File: package-lock.json [hunk 3]
```diff
@@ -6251,7 +5980,8 @@
       }
     },
     "pre-commit": {
-      "version": "github:mongodb-js/pre-commit#f4158768a7ef58b46808cc09a6f6a0994c0baa67",
+      "version": "git+ssh://git@github.com/mongodb-js/pre-commit.git#f4158768a7ef58b46808cc09a6f6a0994c0baa67",
+      "integrity": "sha512-EXGLLZI+SQZOJhWb0EFy0b13fgBwcGQhdXAvbbxMadQYmjv2GQAT2Azhn3unnTOc7zXb45WIoxvfwCZoLq5IOw==",
```
Comment (addaleax) [inline package-lock.json:5984]: If you undo this part of the package-lock.json diff, I think github CI should work. (I don't know why exactly it's failing, though.)

I can also do that for you tomorrow if that's easier.

**Review by addaleax — APPROVED** [review]

**Comment by addaleax** [comment]
@rtritto Awesome, thank you! This is published in `📦  mongodb-query-parser@2.4.9` :+1:

**Comment by rtritto** [comment]
@addaleax if both versions 4 and 5 of `bson` support the legacy option, should the v4 be removed?

**Comment by addaleax** [comment]
@rtritto The reason I kept bson 4 in place for now is that bson 4 and 5 are mutually incompatible (one can’t serialize objects from the other, or at least that’s not officially supported), so if a dependent package wants to use bson 4, it can. I still think that makes sense until it’s reasonable to assume that everyone has migrated to bson 5 (which might be a long time, the Node.js MongoDB driver & bson 5.0.0 releases happened very recently).

**Comment by rtritto** [comment]
@addaleax should `bson` be moved from dependencies to peer dependencies?

**Comment by addaleax** [comment]
Yeah, I think that makes sense :+1:

**Comment by rtritto** [comment]
@addaleax I opened #127

**Pull request merged**